entity-framework/core/testing/index.md

```diff
@@ -1,117 +1,116 @@
 ---
-title: Testing code that uses EF Core - EF Core
+title: Testing applications that use EF Core - EF Core
 description: Different approaches to testing applications that use Entity Framework Core
-author: ajcvickers
-ms.date: 04/22/2020
+author: roji
+ms.date: 11/07/2021
 uid: core/testing/index
 ---
-# Testing code that uses EF Core
+# Testing EF Core Applications
 
-Testing code that accesses a database requires either:
+Testing is an important concern to almost all application types - testing allows you to be sure your application work correctly, and makes it instantly known if its behavior regresses in the future. Since testing may affect how your code is architected, it's highly recommended to plan for a testing early, and to ensure good coverage as your application evolves. This section provides an overview of various testing strategies for applications using EF Core, shows how each one can be implemented and discusses best practices.
 
-* Running queries and updates against the same database system used in production.
-* Running queries and updates against some other easier to manage database system.
-* Using test doubles or some other mechanism to avoid using a database at all.
+## Unit and integration testing
 
-This document outlines the trade-offs involved in each of these choices and shows how EF Core can be used with each approach.
+Although there are many kinds of testing, in this section we'll concentrate on two types of automated testing: unit and integration testing. Without going into formal definitions, unit tests are about isolating a single component and testing against it, with all related components (or dependencies) stubbed out. Importantly, unit tests never interact with a real database, but instead replace the database with a [test double](https://martinfowler.com/bliki/TestDouble.html) which fulfills the same contracts. Integration tests, in contrast, test actual components working together - including your production database system.
 
-> [!TIP]
-> See [EF Core testing sample](xref:core/testing/testing-sample) for code demonstrating the concepts introduced here.
+Integration tests frequently involve the following difficulties:
 
-## All database providers are not equal
+1. In many cases, it's simply not possible or practical to test against the actual external dependency. For example, your application may to interact with some service that cannot be tested again (because of rate limiting, or the lack of a testing environment).
+2. Even when it's possible to involve the real dependency, this may be exceedingly slow: running a large amount of tests against a cloud service may cause tests to take too long. Testing should be part of the developer's everyday workflow, so it's important that tests run quickly.
```

Comment:
More opinionated statements about how/when tests are run.

Reply:
Here I'm trying to list reasons which *may* lead users to abandon the idea of testing against their real database - (speed, isolation...), precisely in order to disprove them later and show that integration testing isn't as bad as they think. I think the "may", "in many cases" make this non-opinionated - just a list of possible reasons why someone would avoid it... 

Any suggestions on making this better?